### PR TITLE
Add support for Slack Events

### PR DESCRIPTION
**Pull request by samgreen96**
This PR adds a Slack Event Handler which can communicate with the [Slack Events API](https://api.slack.com/apis/connections/events-api).

The Slack Events API allows a Slack App to communicate with other Web Services, like the Ruby AU website. For Ruby AU this can be used to send an invitation email to any user who joins the Slack channel. Linked issue: https://github.com/rubyaustralia/ruby_au/issues/131

The basic workflow is as follows:
1. A Slack APP is created, which points to the hook action of the Slack controller `/slack/hook`
2. A challenge is sent by Slack to verify the endpoint
4. The app subscribes to a list of event, in this case the `team_join` event
5. The app is added to the Slack workspace
6. When one of the subscribed to events occurs a request is sent to the endpoint provided in step 1 for processing

I've attempted to create an Event Handler that can be extended in the future, it should be easy to support other events that might be useful.

I have not created a Slack App for Ruby AU, so this would need to be done - probably best to be handled by an admin. Once an App is created we would need to get the token and create an ENV var for it under `SLACK_APP_TOKEN`. I wasn't able to find where these might be configured, so am assuming it's handled within the deployment tool that is being used.

This PR also adds an invitation mailer which is designed to send an invitation email (with a sign up link) to a provided email address. I've merged this functionality with an existing mailer for simplicity. Here's this email:

Screenshot 2023-04-07 at 8 11 06 pm

This wording likely needs to change :) ^ 

I've also added a gem called [letter_opener](https://github.com/ryanb/letter_opener) which opens emails in development in a new tab. This isn't necessary but can make the development experience easier

### CODE REVIEW

File: app/lib/slack/event_handler.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module Slack
+  class EventHandler
+    def self.call(event_data:)
+      event_class = "Slack::Events::#{event_data[:type]&.classify}".constantize
+
+      event_class.call(event_data)
+    rescue NameError, NotImplementedError
+      Rails.logger.error "Unknown Slack event"
+      false
```
Comment (pat) [inline app/lib/slack/event_handler.rb:9]: I wonder if being more clear about the event types we're handling here would be better? Leaning on `classify` and `constantize` and error handling feels a bit risky to me.

An alternative:

```ruby
case event_data[:type]
when "team_join"
  Slack::Events::TeamJoin.call(event_data)
else
  Rails.logger.debug "Unhandled Slack event: #{event_data[:type]}"
end
```

File: app/views/invitation_mailer/invite.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<p>Hi<%= @name && " #{@name}" %>,</p>
+<p>Thanks for joining the Ruby AU Slack Channel. We're super excited for you to join! We'd like to take this opportunity to invite you to become a member of the Ruby Australia website.<p>
+
+<p>If you're interested - <strong>it doesn't cost anything!</strong> - you can sign up below:</p>
+<p class="button"><%= link_to "Join", new_user_registration_url %></p>
+
+<h2>What does being a member mean?</h2>
+<p>As a member, you get to attend and vote at our general meetings, and be nominated (by yourself or others) for a position in the association's committee.</p>
+<p>We will also send you very occasional emails related to the association (perhaps a handful each year).</p>
```
Comment (pat) [inline app/views/invitation_mailer/invite.html.erb:9]: I wonder if we can use the same template for both types of emails here, to ensure we're not doubling up on copy too much? And maybe have a param to determine the context of why someone's being invited.

**Review by pat — COMMENTED** [review]
Thanks so much for this Sam! :) Love that you've tackled a very old issue that's definitely still relevant - the approach you've used looks great to me. I've made a few nitpicking suggestions, but generally a fan of what's here - and that it's tested. 👏🏻

Wrangling the Slack app side of things might take a bit of time - but it'd be good to merge this in sooner rather than later and then deal with that part afterwards. I presume there's nothing stopping us from having this code in production with a placeholder token?

**Comment by samgreen96** [comment]
Thanks for the feedback! All great suggestions which I've gone ahead and changed. We can definitely just use a placeholder token temporarily

**Comment by pat** [comment]
Thanks for the tweaks Sam - I'll merge this in now, and hopefully we can get everything linked up soon.

**Pull request merged**